Explain term automaton queries

MarcusSorealheis opened the pull request:
### Description

This is a PR  to address #12178 that I wrote in my evenings, as I'm OOO on leave for a while and recovering (possibly slow to respond). I'm tapping the open source search community to prevent my edges from dulling and because it's a fun break from less fun stuff. This PR is very likely bad and needs a lot of touch up as I haven't been focused on this work for a while. It may also be a helpful guide for others who are looking to take this issue on and do a much better job. 

I need to update the tests to memorialize the functionality including but not limited to evaluating the explain output for matching docs. I started with handling the null pointer exception due to the origin of my colleague Akaash's issue. 

for context to those unfamiliar, I have mostly worked on janitorial things in this repo, though I am close to some of the adjacent machine learning and developer productivity worlds. I'm a product person not a dev. I volunteer to help open source projects for happiness and rehabilitation. 

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

MarcusSorealheis committed:
start implementing explain for term automaton queries.

MarcusSorealheis committed:
implement Explain for Term Automaton Query and add a test.

MarcusSorealheis committed:
remove cruft.

MarcusSorealheis committed:
add a todo.

MarcusSorealheis committed:
tidying with gradle.

MarcusSorealheis commented:
@rmuir or @mikemccand  Do you have suggestions on who could be a good fit to advise me here on what types of tests or review?

There's a few that are obvious that I need to fix, but I also recognize there maybe some desired strategies around design here. I intend to add more robust tests. This PR touches some important areas of the codebase.

mkhludnev commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -38,23 +38,8 @@
 import org.apache.lucene.index.StoredFields;
 import org.apache.lucene.index.Term;
-import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.*;
```
Sorry. No wildcard imports over there.

MarcusSorealheis replied:
you're right. I know better. In the guide.

MarcusSorealheis replied:
done in [1eede04](https://github.com/apache/lucene/pull/12208/commits/1eede04b8eb3975237e0637fc99963adcf7a1e1f), thanks.

rmuir commented:
For the test question, usually we use QueryUtils which should exercise and test the explanations. https://github.com/apache/lucene/blob/0782535017c9e737350e96fb0f53457c7b8ecf03/lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java

MarcusSorealheis committed:
Update TestTermAutomatonQuery.java

fixing imports

MarcusSorealheis committed:
fixing import goof deleting Explanation.

MarcusSorealheis committed:
test for matching docs.

MarcusSorealheis committed:
removing extra word.

MarcusSorealheis committed:
get clean

MarcusSorealheis commented:
> For the test question, usually we use QueryUtils which should exercise and test the explanations. https://github.com/apache/lucene/blob/0782535017c9e737350e96fb0f53457c7b8ecf03/lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java

Done, in [a828015](https://github.com/apache/lucene/pull/12208/commits/a828015bc9a9b8288317b06a27cd764b10c4a3ac).

I think ready for discussion. It's an iterative improvement. There's endless work, though, and now I feel bad for the maintainers.

MarcusSorealheis committed:
Update TestTermAutomatonQuery.java

i'll update to bson once atlas search is open source. not before. protobufs may be more relevant at this time.

MarcusSorealheis committed:
update Changes.txt.

mikemccand commented:
> I volunteer to help open source projects for happiness and rehabilitation.

Yay :)  I do the same!

> I think ready for discussion. It's an iterative improvement. There's endless work, though, and now I feel bad for the maintainers.

Progress not perfection :)

mikemccand commented:
This is a great start!  I left one small comment.

We can later improve the explanation to state which sequence(s) of term and their positions led to the match.

mikemccand commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonQuery.java:
```diff
@@ -442,8 +442,44 @@ public boolean isCacheable(LeafReaderContext ctx) {
+      }
+
+      Explanation freqExplanation =
+          Explanation.match(score, "TermAutomatonQuery, product of:", termExplanations);
```
Hmm @rmuir noticed that instead of `product of` shouldn't it be `sum of`?

MarcusSorealheis replied:
Yep, this is is an error. Good catch.

`termExplanations.add` 

I believe that maybe my thinking when I wrote this was born out of many operations in the explanation. The score is definitely a sum of products, differences, quotients and more. I'll fix in the next commit.

MarcusSorealheis replied:
fixed in [1a0c77a](https://github.com/apache/lucene/pull/12208/commits/1a0c77a44995ef74025c82f6d84d87cc94b4988c)

MarcusSorealheis committed:
change the txt to be accurate.

zhaih commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonScorer.java:
```diff
@@ -52,6 +52,7 @@ class TermAutomatonScorer extends Scorer {
 
   private int docID = -1;
   private int freq;
+  private EnumAndScorer[] enums;
```
+1

MarcusSorealheis replied:
I intended set the enums field in the constructor, another error. I'll fix that as well.

MarcusSorealheis replied:
Done here: [68c0338](https://github.com/apache/lucene/pull/12208/commits/68c033846beeec815012bfc47e64a011eb3a60b2)

MarcusSorealheis replied:
That could change in future behavior of explain. It's probably better to set the field.

zhaih replied:
I think usually it's better to introduce it when we really need it? But if you insist I'm also OK since it's just another reference.

If you decide to still use a new field then I would suggest:
1. Make it `final`
2. Change the name to `originalSubsOnDoc` or so, since `enums` is a bit general
3. Add some javadoc explaining it is for explain use.

Thanks!

MarcusSorealheis replied:
Makes complete sense. Will get to it shortly.

MarcusSorealheis replied:
Done here: [22ffcd3](https://github.com/apache/lucene/pull/12208/commits/22ffcd34fb0112c48c79516cded1aa0f83972a18)

Do I need to use some annotation or something?

MarcusSorealheis committed:
set the enums field in the constructor.

MarcusSorealheis commented:
> We can later improve the explanation to state which sequence(s) of term and their positions led to the match.

A few things in my hopper but that's going to get done! I would love to do it, but I would love even more if I could mentor a newer engineer on it. In b/t some of my many tasks, maybe I could publish a blog about what a really useful implementation might look like. Of course, explaining it simply feels more difficult than coding it up. 

I once attended a lecture from @ErickErickson on analysis that shaped a bit of my thinking here. I need to pry those notes from where I stored them.

MarcusSorealheis commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonScorer.java:
```diff
@@ -345,6 +347,14 @@ private void countMatches() throws IOException {
+    return enums;
+  }
+
+  public LeafSimScorer getLeafSimScorer() {
```
I believe you're right. Let me check.

MarcusSorealheis replied:
well, you may be right for the other getter, but not this one.

MarcusSorealheis replied:
Can you explain why this should be private?

MarcusSorealheis replied:
Fixed in [29c6fd8](https://github.com/apache/lucene/pull/12208/commits/29c6fd85e648042264dacb27b87d25795997df3f)

MarcusSorealheis replied:
thanks for the review

MarcusSorealheis replied:
I'll let you be the judge. Lol.

MarcusSorealheis committed:
changed to default visibility modifiers.

MarcusSorealheis commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonScorer.java:
```diff
@@ -345,6 +347,14 @@ private void countMatches() throws IOException {
     }
   }
 
+  public EnumAndScorer[] getEnums() {
```
Fixed in [29c6fd8](https://github.com/apache/lucene/pull/12208/commits/29c6fd85e648042264dacb27b87d25795997df3f)

MarcusSorealheis committed:
change name, make final, document.

MarcusSorealheis committed:
tidying up.

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +844,95 @@ public void testRewriteSimplePhrase() throws Exception {
+    TopDocs topDocs = searcher.search(rewrite, 10);
+    assertEquals(0, topDocs.totalHits.value);
+
+    for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
```
Since `totalHits` is 0, shouldn't this loop be skipped? Then if we want to test no match I think we should just feed doc 0 to the searcher?

MarcusSorealheis replied:
the loop is not needed. I added to understand how things were working. I am removing it.

MarcusSorealheis replied:
fixed in : [f91ef20](https://github.com/apache/lucene/pull/12208/commits/f91ef20e5ea706b00c3bf438d9d0b937c146d747)

zhaih replied:
Hey if you put this loop back you're still not exercising explain since there's no document collected.
I think just put `searcher.explain(rewrittenQuery, 0)` should do the trick.

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +844,95 @@ public void testRewriteSimplePhrase() throws Exception {
+    IndexReader r = w.getReader();
+    IndexSearcher searcher = newSearcher(r);
+    Query rewrite = q.rewrite(searcher);
+    assertTrue(rewrite instanceof PhraseQuery);
```
Since we want to test the `TermAutomatonQuery`'s explain, should we try some automaton that will not rewrite to phrase query or multi phrase query?

MarcusSorealheis replied:
This is another error. I should probably test `SynonymGraphFilter.java`

MarcusSorealheis replied:
Which will require some thinking and some set up

MarcusSorealheis replied:
Actually, I need to think about what to do here more. I'm not exactly sure.

MarcusSorealheis replied:
I need to do some digging quickly on the query types and try to understand how to also get the explain query type. I also need to add you to Mike to the changes.txt. You're holding my hand here.

MarcusSorealheis replied:
We should. I made that change in this test, only. 

However, I want to point out that the TermAutomatonQuery's explain will rewrite to a variety of other queries depending on the what the client supplies. I think `PhraseQuery` is merely one of the primitive options when explain is called. The explain is not an exact representation of the true query today. That's something that I think we should look to add because users still benefit from where the work is as of now.

zhaih replied:
> TermAutomatonQuery's explain will rewrite to a variety of other queries

To be more precise, `IndexSearcher` will try to rewrite the query first and then if the rewrite still returns the `TermAutomatonQuery`, then it will create `TermAutomatonWeight` and call the explain you implemented.
So if we want to test the specific implementation (which is what I think we should do here), we need to make sure we created a query that does not rewrite to something else.

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +844,95 @@ public void testRewriteSimplePhrase() throws Exception {
+    IOUtils.close(w, r, dir);
+  }
+
+  // TODO: improve experience of working with explain
```
I do not get what does this TODO means, do you mean making test explain easier?

MarcusSorealheis replied:
I think that explain has many areas of potential improvement. How granular the explain gets, the formatting, testing, among other things.

MarcusSorealheis replied:
it's sort of open ended, tbh. Many things to improve over time.

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +844,95 @@ public void testRewriteSimplePhrase() throws Exception {
+  }
+
+  // TODO: improve experience of working with explain
+
```
Remove empty line?

MarcusSorealheis replied:
good catch

MarcusSorealheis replied:
Fixed in: [f91ef20](https://github.com/apache/lucene/pull/12208/commits/f91ef20e5ea706b00c3bf438d9d0b937c146d747)

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +844,95 @@ public void testRewriteSimplePhrase() throws Exception {
+
+    // Iterate through score docs and assert aspects of explanation
+    for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+      Explanation explanation = searcher.explain(q, scoreDoc.doc);
```
Can we assert somewhere above that searcher rewrite will not change the query type? Since the `searcher.explain` internal will rewrite before it runs explain.

zhaih replied:
Oh I just want to make sure we're testing the target class. 
Since I think this test should test specifically the explain impl of `TermAutomtonQuery` but not `PhraseQuery` or `MultiPhraseQuery` as that should be tested somewhere else.

MarcusSorealheis replied:
I finally think I understand. My approach was to override rewrite to prevent being rewritten to a primitive class. Not sure if that's the right way to approach, so please advise.

zhaih replied:
That's fine I think, thank you!

zhaih reviewed (changes requested):
Thanks for addressing the comments, the logic looks good to me, have some comments on tests.

MarcusSorealheis committed:
change test to term query.

MarcusSorealheis committed:
ensure TAQ is rewritten to appropriate primitive form for explain().

MarcusSorealheis committed:
updated Changes.txt

zhaih reviewed (changes requested):
Sorry this is not what I meant. 
So first, the intend of the added unit test should test the explain code you added, but not rewriting nor explain implementation of TermQuery or PhraseQuery.
Then, look at how IndexSearcher implement explain ([src](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L831)), it will first rewrite the query, and then call the explain for the *rewritten* query's weight. So, if we want to test the code you just added, we need to make sure the rewritten query is actually still TermAutomatonQuery.
Last, let's look at the rewrite logic [here](https://github.com/apache/lucene/blob/main/lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonQuery.java#L529), I haven't dive too deep there but basically there's a comment
```
// Try for either PhraseQuery or MultiPhraseQuery, which only works when the automaton is a
// sausage:
```
So I guess we should be able to figure out a way to let it not rewrite to other query type.

So to summary, I think in the test we should:

1. create a query that won't rewrite to other type
2. call searcher.rewrite and assert the query type to double check
3. check what you claimed (like NoMatchingDocument or 2 documents matched)
4. check the explain implementation, 
  4.1. if there's no document matched, then make sure the explain returns `NO_MATCH` for any documents
  4.2. if there's document matched, make sure the document matched returns proper explanation.

I hope it makes sense

MarcusSorealheis committed:
add a custom extension to TAQ to ensure rewritten query is correct.

MarcusSorealheis committed:
make it easier to redifferentiate terms.

MarcusSorealheis committed:
clean up. How could i forget.

MarcusSorealheis commented:
> Sorry this is not what I meant. So first, the intend of the added unit test should test the explain code you added, but not rewriting nor explain implementation of TermQuery or PhraseQuery. Then, look at how IndexSearcher implement explain ([src](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L831)), it will first rewrite the query, and then call the explain for the _rewritten_ query's weight. So, if we want to test the code you just added, we need to make sure the rewritten query is actually still TermAutomatonQuery. Last, let's look at the rewrite logic [here](https://github.com/apache/lucene/blob/main/lucene/sandbox/src/java/org/apache/lucene/sandbox/search/TermAutomatonQuery.java#L529), I haven't dive too deep there but basically there's a comment
> 
> ```
> // Try for either PhraseQuery or MultiPhraseQuery, which only works when the automaton is a
> // sausage:
> ```
> 
> So I guess we should be able to figure out a way to let it not rewrite to other query type.
> 
> So to summary, I think in the test we should:
> 
> 1. create a query that won't rewrite to other type
> 2. call searcher.rewrite and assert the query type to double check
> 3. check what you claimed (like NoMatchingDocument or 2 documents matched)
> 4. check the explain implementation,
>    4.1. if there's no document matched, then make sure the explain returns `NO_MATCH` for any documents
>    4.2. if there's document matched, make sure the document matched returns proper explanation.
> 
> I hope it makes sense

It does make sense. To be fair, it probably made sense from the first time you said it. I think it needed to percolate in my brain for a bit. I think I've got it now. Or I hope I do. I don't want to get into modifying rewrite because that would be a big change, out of scope for this effort, do the pref testing required a regression risks.

MarcusSorealheis commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +843,91 @@ public void testRewriteSimplePhrase() throws Exception {
     IOUtils.close(w, r, dir);
   }
 
+  /* Implement a custom term automaton query to ensure that rewritten queries
```
@zhaih This felt like the simplest way to force the rewrite to be a `TermAutomatonQuery`

zhaih replied:
I'm ok with it :)

MarcusSorealheis committed:
change accesss modifier and wrap this up.

MarcusSorealheis committed:
needed to add automatic tidy instead of being lazy.

MarcusSorealheis commented:
When I changed the tests to force `TermAutomatonQuery`, I noticed the scores changed. Thanks for that.

zhaih commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +843,92 @@ public void testRewriteSimplePhrase() throws Exception {
+    assertTrue(rewrite instanceof TermAutomatonQuery);
+
+    TopDocs topDocs = searcher.search(rewrite, 10);
+    assertEquals(0, topDocs.totalHits.value);
```
Let's try to call `searcher.explain` here and try to assert it explains a NO_MATCH? otherwise this test case is just testing search?

MarcusSorealheis replied:
I'll do it right now. One second.

MarcusSorealheis replied:
Done in here: [9945f0b](https://github.com/apache/lucene/pull/12208/commits/9945f0b504d1d5c09d4a98f60313b688061f9a1a)

MarcusSorealheis committed:
add test for no match.

MarcusSorealheis committed:
fixing explanation.

MarcusSorealheis commented:
You're right again. Thanks for the help. The tests were the hardest part
and taught me the most.

On Thu, Apr 6, 2023 at 9:17 AM Patrick Zhai ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java
> <https://github.com/apache/lucene/pull/12208#discussion_r1160010214>:
>
> > +
> +    Directory dir = newDirectory();
> +    RandomIndexWriter w = new RandomIndexWriter(random(), dir);
> +    Document doc = new Document();
> +    doc.add(newTextField("field", "protobuf", Field.Store.NO));
> +    w.addDocument(doc);
> +
> +    IndexReader r = w.getReader();
> +    IndexSearcher searcher = newSearcher(r);
> +    Query rewrite = q.rewrite(searcher);
> +    assertTrue(rewrite instanceof PhraseQuery);
> +
> +    TopDocs topDocs = searcher.search(rewrite, 10);
> +    assertEquals(0, topDocs.totalHits.value);
> +
> +    for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
>
> Hey if you put this loop back you're still not exercising explain since
> there's no document collected.
> I think just put searcher.explain(rewrittenQuery, 0) should do the trick.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/lucene/pull/12208#discussion_r1160010214>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAR6TSEISE3TS5WNIENX7F3W73UCLANCNFSM6AAAAAAV6G6DXI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Marcus Eagan

MarcusSorealheis commented:
What I did to test if there was no match was hacky and basically non-functional. Your suggestion is the right way.

zhaih reviewed (approved):
LGTM, thank you!

MarcusSorealheis commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestTermAutomatonQuery.java:
```diff
@@ -842,6 +843,96 @@ public void testRewriteSimplePhrase() throws Exception {
 
+  /* Implement a custom term automaton query to ensure that rewritten queries
+   *  do not get rewritten to primitive queries. The custom extension will allow
+   *  the following explain tests to evaluate Explain.
```
I could add `for the query we need to test, TermAutomatonQuery`

or `TermAutomatonWeights`.

I also think the custom class is so simple, it doesn't need much documenting.

MarcusSorealheis committed:
update the test comment very slightly to be even more clear.

zhaih commented on lucene/CHANGES.txt:
```diff
@@ -78,6 +78,8 @@ Bug Fixes
 * LUCENE-10599: LogMergePolicy is more likely to keep merging segments until
   they reach the maximum merge size. (Adrien Grand)
 
+* GITHUB#12178: Add explanations for TermAutomatonQuery (Marcus Eagan via Patrick Zhai, Mike McCandless, Robert Muir, Mikhail Khludnev)
```
I think we can put it under Lucene 9.6? (it doesn't break any API I think?)

MarcusSorealheis replied:
Good point. Will make the change.

zhaih replied:
I'll just push it to 9x, no worries!

MarcusSorealheis committed:
move the changelog entry.

MarcusSorealheis reviewed (commented):
Made the change.

zhaih committed:
Merge branch 'main' into explain_term_automaton_queries

zhaih commented:
Merged and backported, thank you @MarcusSorealheis
